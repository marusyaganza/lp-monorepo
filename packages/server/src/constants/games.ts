import { Game, GameConfig, Language } from '../generated/graphql';
export const GAMES: GameConfig[] = [
  {
    name: 'WordEcho',
    id: 'audio',
    desc: 'Listen, Type, Master. Sharpen language skills through auditory engagement.',
    type: Game.Audio,
    orderNum: 0,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'LexiSense',
    desc: 'Define & Conquer.Match words with their meanings intuitively.',
    id: 'select_definition',
    type: Game.SelectDef,
    orderNum: 1,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Definify: Match & Learn',
    desc: 'Associate definitions with the right words. A captivating word puzzle adventure!',
    id: 'select_word',
    type: Game.SelectWord,
    orderNum: 2,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'WordRecall: Define & Type',
    desc: 'Type words matching given definitions. Boost vocabulary in engaging rounds!',
    id: 'type_word',
    type: Game.TypeWord,
    orderNum: 3,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'PictureWord Blitz',
    desc: 'Identify objects from pictures and type their names to enhance vocabulary skills!',
    id: 'image',
    type: Game.Image,
    orderNum: 4,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'VoiceWord Challenge',
    desc: 'Speak words from clues and images. Sharpen your vocabulary and speaking skills fast!',
    id: 'speaking',
    type: Game.Speaking,
    orderNum: 5,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Conjugation Quest',
    desc: 'Master verb conjugations across tenses and languages in fast-paced, interactive challenges!',
    id: 'conjugation',
    type: Game.Conjugation,
    orderNum: 6,
    languages: [Language.Spanish]
  },
  {
    name: 'GenderWord Challenge',
    desc: 'Choose nouns and match their gender. Test linguistic skills in thrilling rounds!',
    id: 'gender',
    type: Game.Gender,
    orderNum: 7,
    languages: [Language.Spanish]
  }
];
