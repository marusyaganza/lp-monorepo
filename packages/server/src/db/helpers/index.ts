export * from './sortingHelpers/sortingHelpers';
export * from './general/general';
