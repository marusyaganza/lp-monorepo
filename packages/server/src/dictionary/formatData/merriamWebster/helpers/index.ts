export * from './general/general';
export * from './resource/resource';
export * from './string/string';
export * from './def/def';
