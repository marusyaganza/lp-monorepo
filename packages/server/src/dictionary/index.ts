import { searchWord as searchWordFunc } from './searchWord';
import { mockSearchWord } from './mockSearchWord';

export const searchWord = process.env.USE_MOCKS
  ? mockSearchWord
  : searchWordFunc;
