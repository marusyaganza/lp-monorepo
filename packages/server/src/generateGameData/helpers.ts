import { DEFAULT_TENSE } from '../constants/defaultValues';
import _ from 'lodash';
import { TENSES } from '../constants/tenses';
import {
  DefExample,
  Game,
  Language,
  Tense,
  Word,
  WordDefinition
} from '../generated/graphql';
import { IMinimalWord } from '../types/types';
import { OperationResolutionError } from '../utils/apolloCustomErrors';
import { madeUphWords } from '../constants/madeUpWord';

export function getExamples(defs?: (WordDefinition | null)[] | null) {
  if (!Array.isArray(defs)) {
    return;
  }

  const examples = defs
    .flatMap(def => def?.examples)
    .filter(Boolean)
    .slice(0, 5);

  return examples as DefExample[];
}

export function getGameTask(gameType: Game, tenseInput?: Tense | null) {
  const tense = tenseInput || DEFAULT_TENSE;
  const tasks = {
    [Game.Audio]: "Type the word that you've heard",
    [Game.SelectDef]: 'Select a definition that means ',
    [Game.SelectWord]: 'Select a word that means ',
    [Game.TypeWord]: 'Type a word that means ',
    [Game.Image]: 'Type the name of the object shown in the picture.',
    [Game.Gender]: 'Pick the Correct Gender',
    [Game.Speaking]:
      'Click the microphone and say the word matching the description.',
    [Game.Conjugation]: `Conjugate the verb in ${TENSES[tense]}`
  };
  return tasks[gameType];
}

/**Replaces word's `name` in `def` with `[...]` block
 * required to exclude the answer from the question
 */
export function prepareDef(def?: string | null, name?: string | null): string {
  if (!def || !name) {
    return '';
  }

  const replacement = ' [...] ';
  const endReplacement = ' [...]';
  const word = name.toLocaleLowerCase();

  let result = def.replaceAll(new RegExp(` ${word} `, 'gi'), replacement);

  if (result.toLocaleLowerCase().startsWith(`${word} `)) {
    result = result.replace(new RegExp(`${word} `, 'i'), replacement).trim();
  }

  if (
    result
      .toLocaleLowerCase()
      .replace(/[.,#!$%^&*;:{}=\-_`~()]/g, '')
      .endsWith(` ${word}`)
  ) {
    result = result.replace(new RegExp(` ${word}`, 'i'), endReplacement).trim();
  }

  return result;
}

export function generateOptions(
  data: Word[],
  count: number,
  currentWord: Word,
  language: Language
) {
  const extraWords = madeUphWords[language];

  const optCandidates: IMinimalWord[] = data.filter(
    word => word.id !== currentWord.id
  );

  let options: IMinimalWord[];

  if (optCandidates.length + extraWords.length < count - 1) {
    throw new OperationResolutionError(`Not enough words to generate options.`);
  }

  if (optCandidates.length < count) {
    const moreOptions = _.take(extraWords, count - optCandidates.length - 1);
    options = [...optCandidates, ...moreOptions];
  } else {
    const mixedWords = _.shuffle(optCandidates);
    options = _.take(mixedWords, count - 1);
  }

  options = [...options, currentWord];

  return _.shuffle(options);
}
