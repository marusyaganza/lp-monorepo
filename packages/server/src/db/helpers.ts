import { Document, Types } from 'mongoose';
import {
  Maybe,
  Language,
  Game,
  SortBy,
  SortWordsBy
} from '../generated/graphql';

type ModelDataType = {
  id?: Maybe<string>;
  user?: string;
};

interface DbData extends ModelDataType {
  _id?: string;
}

type DocumentType<T> =
  | (Document<unknown, unknown, T> &
      T & {
        _id: Types.ObjectId;
      })
  | null;

export interface WordsFilterType {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder?: boolean;
  timesToLearn?: number | null;
  gameType?: Game;
  user?: string;
  tags?: string[];
  searchQuery?: string;
}

export interface WordsWithPaginationFilter extends WordsFilterType {
  page?: number;
  limit?: number;
}

export function formatData<T>(data: DocumentType<T>): T | null {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter?.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
const NO_TAGS_ID = '000000000000000000000000';

export function getWordsFilters(filter: WordsFilterType) {
  const {
    user,
    language = Language.English,
    sortBy,
    isReverseOrder,
    gameType,
    tags,
    searchQuery,
    timesToLearn = 5
  } = filter;

  // filter for a specific game
  let gameFilter;
  if (gameType === Game.Conjugation) {
    gameFilter = { conjugation: { $ne: null } };
  }
  if (gameType === Game.Audio) {
    gameFilter = { audioUrl: { $ne: '' } };
  }

  // filter based on tags
  let tagsFilters: { tags: string }[] = [];
  if (tags?.length) {
    tagsFilters = tags
      ?.filter(tag => tag !== NO_TAGS_ID)
      .map(tag => {
        return { tags: tag };
      });
  }

  let noTagsFilter;
  if (tags?.includes(NO_TAGS_ID)) {
    noTagsFilter = { $or: [{ tags: [] }, { tags: null }] };
  }

  let searchFilter;
  if (searchQuery) {
    const searchRegexp = { $regex: new RegExp(searchQuery, 'i') };
    searchFilter = {
      $or: [
        { name: searchRegexp },
        { 'defs.def': searchRegexp },
        { alternativeSpelling: searchRegexp },
        { stems: searchRegexp }
      ]
    };
  }

  // filter based on learning status
  let learningStatusFilters: Record<string, unknown>[] = [];
  if (gameType && sortBy !== SortBy.MemoryRefresher) {
    learningStatusFilters = [
      { isLearned: { $ne: true } },
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
          { [`statistics.${gameType}.successRate`]: null }
        ]
      }
    ];
  }
  if (sortBy === SortBy.MemoryRefresher) {
    learningStatusFilters = [
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
          { isLearned: true }
        ]
      }
    ];
  }

  //sorting
  let orderNum = isReverseOrder ? -1 : 1;

  // words with the highest number of errors should go first
  if (sortBy === SortBy.ErrorCount) {
    orderNum = -1 * orderNum;
  }

  let sort: Record<string, number> = { $natural: orderNum };

  if (sortBy) {
    let propName: string = sortBy;
    if (gameType) {
      propName = `statistics.${gameType}.${sortBy}`;
    }

    if (sortBy !== SortBy.MemoryRefresher) {
      sort = { [propName]: orderNum };
    }
  }

  const filtersArray = [
    gameFilter,
    noTagsFilter,
    searchFilter,
    ...tagsFilters,
    ...learningStatusFilters
  ].filter(Boolean);

  const filters: Record<string, unknown> = {
    user,
    language
  };

  if (filtersArray?.length) {
    filters.$and = filtersArray;
  }

  return {
    sort,
    filters
  };
}
