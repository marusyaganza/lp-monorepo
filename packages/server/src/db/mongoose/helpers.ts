import { Types, Document } from 'mongoose';
import { User, Word } from '../../generated/graphql';
import { DbData } from 'db';

type DocumentType =
  | (Document<unknown, Record<string, any>, User | Word> &
      Omit<
        | User
        | (Word & {
            _id: Types.ObjectId;
          }),
        never
      >)
  | null;

export function formatData(data: DocumentType) {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
