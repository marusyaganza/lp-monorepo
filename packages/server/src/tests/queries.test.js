import { createTestServer } from './helpers';
import { words, users } from './mocks/data';
import { Language, Game, SortBy, SortWordsBy } from '../generated/graphql';
import { searchWord } from '../dictionary';
import { games } from '../mocks/games';
import { generateGameData } from '../utils/generateGameData';
import {
  randomNumbersArray,
  words as wordsArr
} from './mocks/gameGenerationData';
const wordsQuery = `
  {
    words {
      id
      name
      defs {
        def
      }
      particle
      audioUrl
      transcription
      isOffensive
      level
    }
  }
`;

const wordByIdQuery = `
  query Query($wordId: ID!) {
    word(id: $wordId) {
      id
      name
      defs {
        def
        examples {
          text
        }
      }
      particle
      imgUrl
      audioUrl
      additionalInfo
      transcription
      user
      isOffensive
      stems
      level
    }
  }
`;

const userQuery = `
  {
    user {
      firstName
      lastName
      email
      id
      role
    }
  }
`;

const searchQuery = `
  query SearchWords($input: WordSearchInput!) {
    searchWord(input: $input) {
      ... on Suggestions {
        suggestions
      }
      ... on DictionaryWord {
        uuid
        transcription
        stems
        particle
        name
        isOffensive
        imgUrl
        defs {
          examples {
            text
            translation
          }
          def
        }
        audioUrl
        additionalInfo
        imgDesc
        shortDef
      }
    }
  }
`;

const gamesQuery = `query Games {
  games {
    desc
    imgUrl
    name
    id
  }
}`;

const gameQuery = `query Game($input: GameDataInput!) {
  game(input: $input) {
    questions {
      wordId
      question
      options
      answer
      additionalInfo {
        audioUrl
      }
    }
    task
    type
  }
}`;

let index = 0;

const mathRandomMock = jest
  .spyOn(global.Math, 'random')
  .mockImplementation(function () {
    if (index === randomNumbersArray.length) {
      index = 0;
    }
    const randomNum = randomNumbersArray[index];
    index++;
    return randomNum;
  });

const findMany = jest.fn(() => words);
const findOneUser = jest.fn(() => users[0]);
const findOneWord = jest.fn(() => words[0]);
const findManyAndSort = jest.fn(() => wordsArr);

const user = { id: '1' };

const models = {
  Word: {
    findMany,
    findOne: findOneWord,
    findManyAndSort
  },
  User: {
    findOne: findOneUser
  }
};

describe('queries', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });
  test('words', async () => {
    const { query } = createTestServer({
      user,
      models
    });

    const res = await query({
      query: wordsQuery
    });
    expect(res).toMatchSnapshot();
    expect(findManyAndSort).toHaveBeenCalledTimes(1);
  });

  test('word by id', async () => {
    const { query } = createTestServer({
      user,
      models
    });

    const res = await query({
      query: wordByIdQuery,
      variables: { wordId: 'mockid' }
    });
    expect(res).toMatchSnapshot();
    expect(findOneWord).toHaveBeenCalledTimes(1);
  });

  test('word by id if word is not found', async () => {
    const { query } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          findOne: jest.fn()
        }
      }
    });

    const res = await query({
      query: wordByIdQuery,
      variables: { wordId: 'mockid' }
    });
    expect(res).toMatchSnapshot();
  });

  test('user if user exist', async () => {
    const { query } = createTestServer({
      user: { id: '1', role: 'MEMBER' },
      models
    });

    const res = await query({ query: userQuery });
    expect(res).toMatchSnapshot();
  });

  test('user if user is undefined', async () => {
    const { query } = createTestServer({
      models
    });

    const res = await query({ query: userQuery });
    expect(res).toMatchSnapshot();
  });

  test('user if user is not found', async () => {
    const { query } = createTestServer({
      user: { id: '1', role: 'MEMBER' },
      models: {
        ...models,
        User: {
          findOne: jest.fn()
        }
      }
    });

    const res = await query({ query: userQuery });
    expect(res).toMatchSnapshot();
  });

  test('searchWord', async () => {
    const searchWord = jest.fn(() => words);
    const { query } = createTestServer({
      user,
      searchWord,
      models
    });
    await query({
      query: searchQuery,
      variables: {
        input: {
          search: 'word',
          language: Language.English
        }
      }
    });
    expect(searchWord).toHaveBeenCalledWith('word', Language.English);
  });

  test('searchWord spanish lang, mocks enabled', async () => {
    const { query } = createTestServer({
      user,
      searchWord,
      models
    });
    const res = await query({
      query: searchQuery,
      variables: {
        input: {
          search: 'idioma',
          language: Language.Spanish
        }
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('searchWord english lang, mocks enabled', async () => {
    const { query } = createTestServer({
      user,
      searchWord,
      models
    });
    const res = await query({
      query: searchQuery,
      variables: {
        input: {
          search: 'rubber',
          language: Language.English
        }
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('searchWord spanish lang, mocks enabled returns suggestion', async () => {
    const { query } = createTestServer({
      user,
      searchWord,
      models
    });
    const res = await query({
      query: searchQuery,
      variables: {
        input: {
          search: 'unbelievableWordThatDoNotExist',
          language: Language.Spanish
        }
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('searchWord english lang, mocks enabled returns suggestion', async () => {
    const { query } = createTestServer({
      user,
      searchWord,
      models
    });
    const res = await query({
      query: searchQuery,
      variables: {
        input: {
          search: 'unbelievableWordThatDoNotExist',
          language: Language.English
        }
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('games', async () => {
    const { query } = createTestServer({
      games
    });
    const res = await query({ query: gamesQuery });
    expect(res).toMatchSnapshot();
  });

  test('get gamedata with walid user for Audio game and English', async () => {
    const { query } = createTestServer({
      games,
      generateGameData,
      models,
      user
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input: { gameType: Game.Audio, language: Language.English }
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('get gamedata with walid user for SelectDef game and English', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.PracticedTimes
    };
    const { query } = createTestServer({
      games,
      generateGameData,
      models,
      user
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(findManyAndSort).toHaveBeenCalledWith({
      ...input,
      user: user.id
    });
  });

  test('get gamedata with walid user for SelectWord game and English', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: false,
      sortBy: SortBy.LastTimePracticed
    };

    const { query } = createTestServer({
      games,
      generateGameData,
      models,
      user
    });

    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(findManyAndSort).toHaveBeenCalledWith({
      ...input,
      user: user.id
    });
  });

  test('get gamedata with walid user for TypeWord game and English', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      games,
      generateGameData,
      models,
      user
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(findManyAndSort).toHaveBeenCalledWith({
      ...input,
      user: user.id
    });
  });

  test('should not generate game data if user is not found', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      games,
      generateGameData,
      models: {
        ...models,
        User: {
          findOne: jest.fn()
        }
      },
      user
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(findManyAndSort).toHaveBeenCalledTimes(0);
  });

  test('should not generate game data if user is undefined', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      games,
      generateGameData,
      models
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(findManyAndSort).toHaveBeenCalledTimes(0);
  });

  test('generate game data if user does not have enough words', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      games,
      generateGameData,
      user,
      models: {
        ...models,
        Word: {
          findManyAndSort: jest.fn(() => [])
        }
      }
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('generate game data if words do not exist', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      games,
      generateGameData,
      user,
      models: {
        ...models,
        Word: {
          findManyAndSort: jest.fn()
        }
      }
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('generate game data if game is not found', async () => {
    const input = {
      gameType: Game.SelectDef,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      games: [],
      generateGameData,
      user,
      models: {
        ...models,
        Word: {
          findManyAndSort: jest.fn()
        }
      }
    });
    const res = await query({
      query: gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
  });
});
