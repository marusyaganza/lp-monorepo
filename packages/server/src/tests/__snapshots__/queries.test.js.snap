// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries searchWord english lang, mocks enabled 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "searchWord": [
          {
            "additionalInfo": null,
            "audioUrl": "mock_audio_endpoint/en/us/mp3/r/rubber01.mp3",
            "defs": [
              {
                "def": "{dx}see also {dxt|foam rubber||}{/dx}",
                "examples": [
                  "tires made of {it}rubber{/it}",
                ],
              },
              {
                "def": "{bc}{sx|eraser||}",
                "examples": null,
              },
              {
                "def": "{bc}shoes or boots that are made of rubber and that fit over your regular shoes to keep them dry {dx}compare {dxt|galoshes||}{/dx}",
                "examples": null,
              },
              {
                "def": "{bc}{sx|condom||}",
                "examples": null,
              },
              {
                "def": "{bc}a flat piece of hard, white rubber on which the pitcher stands while pitching",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber",
            "particle": "noun",
            "stems": [
              "rubber",
              "rubbers",
              "burn rubber",
            ],
            "transcription": "rub*ber",
            "uuid": "59e03286-73c9-43e6-b1eb-4407a6c06b1e",
          },
          {
            "additionalInfo": null,
            "audioUrl": null,
            "defs": [
              {
                "def": "{bc}a contest that consists of a series of games and that is won by the player or team that wins the most games (such as two games out of three) ",
                "examples": [
                  "Let's play a {it}rubber{/it} of bridge.",
                ],
              },
              {
                "def": "{bc}a game that decides who will be the winner of a series of games",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber",
            "particle": "noun",
            "stems": [
              "rubber",
              "caoutchouc",
              "india rubber",
              "rubbers",
              "rubber game",
            ],
            "transcription": "rubber",
            "uuid": "31c81e56-7a47-40ee-b6b3-63407105feb4",
          },
          {
            "additionalInfo": null,
            "audioUrl": "mock_audio_endpoint/en/us/mp3/c/condom01.mp3",
            "defs": [
              {
                "def": "{bc}a thin rubber covering that a man wears on his penis during sex in order to prevent a woman from becoming pregnant or to prevent the spread of diseases ",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "condom",
            "particle": "noun",
            "stems": [
              "condom",
              "condoms",
              "rubber",
              "sheath",
            ],
            "transcription": "con*dom",
            "uuid": "79b4d0fe-8d9d-4bd0-879c-366078d586d6",
          },
          {
            "additionalInfo": null,
            "audioUrl": "mock_audio_endpoint/en/us/mp3/e/eraser01.mp3",
            "defs": [
              {
                "def": "{dx}see picture at {dxt|office||}{/dx}",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "eraser",
            "particle": "noun",
            "stems": [
              "eraser",
              "erasers",
              "rubber",
            ],
            "transcription": "eras*er",
            "uuid": "eb991c0d-a218-4812-930d-9ea376c44445",
          },
          {
            "additionalInfo": null,
            "audioUrl": null,
            "defs": [
              {
                "def": "{dx}see picture at {dxt|office||}{/dx}",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber band",
            "particle": "noun",
            "stems": [
              "rubber band",
              "rubber bands",
              "elastic",
              "elastic band",
            ],
            "transcription": "rubber band",
            "uuid": "0634ec1c-dc66-465f-b390-fb945313d066",
          },
          {
            "additionalInfo": null,
            "audioUrl": null,
            "defs": [
              {
                "def": "{bc}a tall boot that is made out of rubber and that keeps your feet and lower legs dry ",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber boot",
            "particle": "noun",
            "stems": [
              "rubber boot",
              "rubber boots",
              "wellington",
            ],
            "transcription": "rubber boot",
            "uuid": "48d907a0-a8fb-4ca3-bc75-57a70d908166",
          },
          {
            "additionalInfo": null,
            "audioUrl": null,
            "defs": [
              {
                "def": "{bc}a bullet that is made out of rubber so that it hurts but does not kill people and that is used by the police and military to control a crowd",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber bullet",
            "particle": "noun",
            "stems": [
              "rubber bullet",
              "rubber bullets",
            ],
            "transcription": "rubber bullet",
            "uuid": "dbdc0d78-3e16-4dc3-afd5-2e72bf8b6a55",
          },
          {
            "additionalInfo": null,
            "audioUrl": null,
            "defs": [
              {
                "def": "{bc}a thick, stretchy glue that contains rubber",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber cement",
            "particle": "noun",
            "stems": [
              "rubber cement",
              "rubber cements",
            ],
            "transcription": "rubber cement",
            "uuid": "eb1c2aba-f3df-46d5-936b-626e499998b1",
          },
          {
            "additionalInfo": null,
            "audioUrl": null,
            "defs": [
              {
                "def": "{bc}a plant with large, shiny, dark green leaves that is often grown indoors",
                "examples": null,
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber plant",
            "particle": "noun",
            "stems": [
              "rubber plant",
              "rubber plants",
            ],
            "transcription": "rubber plant",
            "uuid": "39f6a4ec-4959-4d6f-876a-39b8f780c8cd",
          },
          {
            "additionalInfo": null,
            "audioUrl": null,
            "defs": [
              {
                "def": "{bc}a small tool that consists of a handle attached to a piece of carved rubber and that is dipped in ink and then pressed on paper to print a date, name, symbol, etc.",
                "examples": null,
              },
              {
                "def": "{bc}a person or organization that automatically approves everything that someone does or decides ",
                "examples": [
                  "The legislature has been nothing more than a {it}rubber stamp{/it} for the President.",
                ],
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "rubber stamp",
            "particle": "noun",
            "stems": [
              "rubber stamp",
              "rubber stamps",
              "rubber-stamp",
            ],
            "transcription": "rubber stamp",
            "uuid": "3d83c22d-0913-4b18-8462-a4064a7d8fd2",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries searchWord english lang, mocks enabled returns suggestion 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "searchWord": [
          {
            "suggestions": [
              "Mocking is enabled. You can query words from the list: egalitarian, rubber, pussy, heart, wheel voluminous",
            ],
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries searchWord spanish lang, mocks enabled 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "searchWord": [
          {
            "additionalInfo": null,
            "audioUrl": "mock_audio_endpoint/es/me/mp3/i/idiom01sp.mp3",
            "defs": [
              {
                "def": "{bc}{a_link|language} ",
                "examples": [
                  "el idioma ingl√©s",
                ],
              },
            ],
            "imgDesc": null,
            "imgUrl": null,
            "isOffensive": false,
            "name": "idioma",
            "particle": "masculine noun",
            "stems": [
              "idioma",
            ],
            "transcription": "idioma",
            "uuid": "63882a81-032e-4dc1-9c22-bb683cb1bad9",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries searchWord spanish lang, mocks enabled returns suggestion 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "searchWord": [
          {
            "suggestions": [
              "Mocking is enabled. You can query words from the list: idioma",
            ],
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries user 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "user": {
          "email": "member@member.com",
          "firstName": "User",
          "id": "1",
          "lastName": "Test",
          "role": "MEMBER",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries word by id 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "word": {
          "additionalInfo": null,
          "audioUrl": "https://www.oxfordlearnersdictionaries.com/media/english/uk_pron/e/ega/egali/egalitarian__gb_1.mp3",
          "defs": [
            {
              "def": "a person who believes that everyone is equal and should have the same rights and opportunities",
              "examples": [
                "He described himself as ‚Äòan {it}egalitarian{/it}‚Äô.",
              ],
            },
            {
              "def": "aiming for equal wealth, status, etc., for all people",
              "examples": [
                "He is a committed {it}egalitarian{/it}",
                "Even in the {it}egalitarian{/it} Nordic country, Marin felt hergender and age sometimes received too much emphasis. Kostya Manenkov And Karl Ritter, USA TODAY, 21 Aug. 2022",
              ],
            },
          ],
          "id": "mockid",
          "imgUrl": null,
          "isOffensive": false,
          "level": "C1",
          "name": "egalitarian",
          "particle": "noun",
          "stems": [
            "egalitarian",
            "egalitarianism",
            "egalitarians",
          ],
          "transcription": "iÀå…°√¶l…™Ààte…ôri…ôn",
          "user": "1",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries words 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "words": [
          {
            "audioUrl": "https://www.oxfordlearnersdictionaries.com/media/english/uk_pron/e/ega/egali/egalitarian__gb_1.mp3",
            "defs": [
              {
                "def": "a person who believes that everyone is equal and should have the same rights and opportunities",
              },
              {
                "def": "aiming for equal wealth, status, etc., for all people",
              },
            ],
            "id": "mockid",
            "isOffensive": false,
            "level": "C1",
            "name": "egalitarian",
            "particle": "noun",
            "transcription": "iÀå…°√¶l…™Ààte…ôri…ôn",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
